blog: lw7 o11y day (#13451)
supabase/supabase: apps/www/_blog/2023-04-10-o11y-day.mdx

Logflare is a multi-node, highly available Elixir cluster, ingesting the log events and storing them into BigQuery for Supabase and Logflare’s customers. On average, the cluster has 6 nodes handling every spike our customers throw at it.

To expose log data to customers, we leverage Logflare Endpoints. This provides an HTTP integration into Supabase Studio, powering the log query UIs and most timeseries charts. These charts live across the studio, such as the project home page and the new API reports.

{/* <!--  add chart screenshot --> */}

## Self-hosting Logflare

Logflare was available under a BSL license prior to joining Supabase. We’ve since changed the license to [Apache 2.0](https://github.com/Logflare/logflare/blob/staging/LICENSE.md), aligning it with our open source philosophy.

In the past few months we’ve made Logflare more developer-friendly for local development and self-hosting. While you’re building a project, you can view and query your logs from any Supabase service, just as you would in our cloud Platform.

![Realtime local logs](/images/blog/lw7-o11y/realtime-logs-screenshot.png)

<div className="bg-gray-300 rounded-lg px-6 py-2 italic">

> 📢 Check out the [new self-hosting docs](http://localhost:3001/docs/reference/self-hosting-analytics/introduction) to get Logflare up and running as your analytics server.

Reviewer: Edited URL

```suggestion
📢 Check out the [new self-hosting docs](https://supabase.com/docs/reference/self-hosting-analytics/introduction) to get Logflare up and running as your analytics server.
```

With Endpoints, you can create HTTP API endpoints from any SQL query, including parameterised queries. Endpoints is a bit like a PostgREST view but with some benefits:

- Query parameters
  - Query parameters of your Endpoint query URL are provided as query string parameters to the SQL query.
- Read-through caching
  - Results from the query are cached in memory for fast response times.
  - A read-through cache provides results if cached results do not exist.
- Active cache warming
  - Query results are proactively warmed at a configurable interval for a combination of fast response times and as-realtime-as-needed data.
- Query sandboxing
  - If an Endpoint query contains a CTE and the sandbox option is selected, the Endpoint will inject the query string of the `sql` query parameter into the Endpoint SQL replacing the default query (the part of the SQL query after the CTE).
  - Endpoints parse SQL to allow `select` queries only. No DML or DDL statements are permitted to run through Logflare Endpoints.

> This feature set has let Supabase build any view we’ve needed on top of billions of log events processed daily.

Reviewer 2: ```suggestion
With this feature set, Supabase has been able to build any view we’ve needed on top of billions of daily log events.
```

### Logflare Endpoint Example

> This example showcases a GitHub webhook that sends all event in all Supabase GitHub organizations to Logflare. This gives Logflare structured events where the payload of the Github webhook gets turned into metadata for a log event:

Reviewer 2: ```suggestion
Using webhooks, we can send all GitHub events in the Supabase organization to Logflare. The webhook sends structured events, and Logflare transforms the payload into metadata:
```

```json
{
  "event_message": "supabase/supabase | JohannesBauer97 | created",
  "id": "0d48b71d-91c5-4356-82c7-fdb299b625d0",
  "metadata": {
    "sender": {
      "id": 15695124,
      "login": "JohannesBauer97",
      "node_id": "MDQ6VXNlcjE1Njk1MTI0",
      "site_admin": false,
      "type": "User",
      "url": "https://api.github.com/users/JohannesBauer97"
    },

>     "starred_at": "2023-03-30T20:33:55Z",
> 		...
>   },

Reviewer 2: ```suggestion
    "starred_at": "2023-03-30T20:33:55Z",
    ...
  },
```

  "timestamp": 1680208436849642
}
```

> We’re interested in the top contributors org wide, and this can be extracted by creating an Endpoint with SQL (in BigQuery dialect):

Reviewer 2: ```suggestion
We’re interested in the top contributors, which can be extracted with SQL (in BigQuery dialect):
```

```sql
select
  count(t.timestamp) as count,
  s.login as gh_user
from
  `github.supabase.webhooks` as t
  cross join unnest(metadata) as m
  cross join unnest(m.sender) as s
where
  timestamp::date > current_date() - @day::int
group by
  gh_user
order by
  count desc
limit
  25
```

With this view in place, we can use Endpoints to provide an API that we can hit from our application:

```bash
curl "https://logflare.app/endpoints/query/69425db0-1cfb-48b4-84c7-2a872b6f0a61" \
 -H 'Content-Type: application/json; charset=utf-8' \
 -G -d "day=30"
```

> Which gives us the response with the top org wide contributors for the last 30 days!

Reviewer 2: ```suggestion
This returns a JSON response with the top org wide contributors for the last 30 days!
```

```json
{
  "result": [
    {
      "count": 23404,
      "gh_user": "vercel[bot]"
    },
    {
      "count": 10005,
      "gh_user": "joshenlim"
    },
    {
      "count": 7026,
      "gh_user": "MildTomato"
    },
    {
      "count": 6405,
      "gh_user": "fsansalvadore"
    },
    {
      "count": 5195,
      "gh_user": "saltcod"
    },
    {
      "count": 3454,
      "gh_user": "alaister"
    },
    {
      "count": 2691,
      "gh_user": "kevcodez"
    },
    {
      "count": 2117,
      "gh_user": "gregnr"
    },
    {
      "count": 1769,
      "gh_user": "Ziinc"
    },
    {
      "count": 1749,
      "gh_user": "chasers"
    },
    {
      "count": 1430,
      "gh_user": "Isaiah-Hamilton"
    },
    ...
  ]
}
```

> We can then configure this Endpoint to cache results for a set interval of 10 minutes after the first API request, and proactively update those cached results every 2 minutes. This would result in 5 separate queries spread out áo cross the 10 minute cache time. Even if we were to throw tens of thousands of API requests at the endpoint, we would only sustain the cost of 5 queries to update the cached data.

Reviewer 2: ```suggestion
We can configure this Endpoint to cache results for an interval of 10 minutes after the first API request, and proactively update those cached results every 2 minutes - 5 queries across the 10 minute interval. Even if we hit the Endpoint thousands of times, we only sustain the cost of 5 queries.
```

> The initial request is fast because Logflare also performs setup (such as partitioning) on our BigQuery tables appropriately. Subsequent requests made are extremely fast as they get cached in-memory.

Reviewer 2: ```suggestion
The initial request is fast because Logflare also performs setup (such as partitioning) on our BigQuery tables appropriately. Subsequent requests are _extremely fast_ as they are cached in-memory.
```

> The best part is that all these knobs can be tweaked based on the use case. If we have a hard realtime requirement, we can completely disable caching, or drop the proactive caching to update on a per-second basis.

Reviewer 2: ```suggestion
The best part is that all these knobs can be tweaked for your use case. If we have a real-time requirement, we can completely disable caching or reduce the proactive caching to update on a per-second interval.
```

## The Self-hosted Challenge

> In order to execute the license change, we had to get rid of all closed-source dependencies. Prior to our efforts, Logflare relied on the closed source [General SQL Parser](https://www.sqlparser.com/) under a business licenses. However, it would not be an acceptable compromise under the Apache License.

Reviewer 2: ```suggestion
To change the license, we needed to remove all closed-source dependencies. Previously, Logflare relied on the closed source [General SQL Parser](https://www.sqlparser.com/) under a business licenses. This is incompatible with the Apache License.
```

> This resulted in an effort to switch Logflare over to an open source alternative, and we settled on the rust-based [sqlparser-rs](https://github.com/sqlparser-rs/sqlparser-rs) library, and also ended up helping to contribute a [few fixes](https://github.com/sqlparser-rs/sqlparser-rs/pulls?q=is%3Apr+is%3Amerged+author%3AZiinc) upstream for the BigQuery dialect.

Reviewer 2: ```suggestion
We switched to an open source alternative, the rust-based [sqlparser-rs](https://github.com/sqlparser-rs/sqlparser-rs) library, and contributed a [few updates](https://github.com/sqlparser-rs/sqlparser-rs/pulls?q=is%3Apr+is%3Amerged+author%3AZiinc) for the BigQuery dialect.
```

> Besides the parser, a big part of getting Logflare working with Supabase locally was transforming the multi-tenant architecture into something that was self-hosting friendly and could be easily configurable. We also moved towards environment variable based configuration instead of compile-time configurations, and also exposed the Endpoints configurations necessary for Supabase Logs.

Reviewer 2: ```suggestion
Along with the parser, we invested a lot of effort into transforming the multi-tenant architecture into something that was self-hosting friendly and easily configurable. We moved towards environment variable based configuration instead of compile-time configurations, exposing the Endpoints configurations necessary for Supabase Logs.
```

## What’s Next?

To further integrate Logflare into the Supabase platform, we are building out 2 main areas: Management API, Multiple Backends.

### Management API

> The Management API allows users to programmatically interact with Logflare server and manage their Logflare account and resources. This feature will be available for both Logflare customers and self-hosted users.

Reviewer 2: ```suggestion
The Management API allows users to interact programmatically with Logflare to manage their account and resources. This feature will be available for both Logflare customers and self-hosted users.
```

You can check out the preview of our OpenAPI spec here: [https://logflare.app/swaggerui](https://logflare.app/swaggerui)

![Swagger UI](/images/blog/lw7-o11y/swagger.png)

> Not only that, we intend to expose user account provisioning to select partners. Soon, you’ll be able to apply to become a Logflare Partner and provision Logflare accounts with the Partner API. Perfect if you want to resell a log analytics service from within your platform.

Reviewer 2: ```suggestion
Not only that, we intend to expose user account provisioning to select partners. Soon, you’ll be able to become a Logflare Partner to provision Logflare accounts through the Partner API. Perfect if you want to resell a log analytics service from your own platform.
```

Contact us at growth@supabase.com to get in early on that waitlist.

## Multiple Backends

> We understand that many of our customers are concerned about relying on closed-source products. Although Logflare currently supports a BigQuery backend, we hope to add support for other analytics optimized databases like Clickhouse. Beyond that, we can also support pushing data to other web services, allowing Logflare to fit into any data pipeline.

Reviewer 2: ```suggestion
Logflare currently supports a BigQuery backend. We plan to add support for other analytics-optimized databases, like Clickhouse. We will also support pushing data to other web services, making Logflare a good fit for any data pipeline.
```

> This would benefit our Supabase CLI integration, such that once Postgres support is available, Logflare would be able to integrate seamlessly out-of-the-box without the BigQuery requirement.

Reviewer 2: ```suggestion
This will benefit the Supabase CLI: once Postgres support is available, Logflare will be able to integrate seamlessly, without the BigQuery requirement.
```

---

> title: Supabase Logs are now self-hosted

Reviewer 3: ```suggestion
title: Supabase Logs are now self-hosted
launchweek: 7
```

Reviewer 3: @Ziinc Can you please just add this `launchweek: 7` to the meta data?
We'll use it to show some additional links at the end of the post.

Reviewer 4: done

---
title: Supabase Logs are now self-hosted
launchweek: 7
tags:
  - launch-week
  - o11y
date: '2023-04-10'
to_depth: 3
author: chasers,ziinc
---

> Today, we’re releasing Supabase Logs for both self-hosted users and CLI development

Reviewer 2: ```suggestion
Today, we’re releasing Supabase Logs for both self-hosted users and CLI development.
```

Logflare is a multi-node, highly available Elixir cluster, ingesting the log events and storing them into BigQuery for Supabase and Logflare’s customers. On average, the cluster has 6 nodes handling every spike our customers throw at it.

To expose log data to customers, we leverage Logflare Endpoints. This provides an HTTP integration into Supabase Studio, powering the log query UIs and most timeseries charts. These charts live across the studio, such as the project home page and the new API reports.

{/* <!--  add chart screenshot --> */}

## Self-hosting Logflare

Logflare was available under a BSL license prior to joining Supabase. We’ve since changed the license to [Apache 2.0](https://github.com/Logflare/logflare/blob/staging/LICENSE.md), aligning it with our open source philosophy.

In the past few months we’ve made Logflare more developer-friendly for local development and self-hosting. While you’re building a project, you can view and query your logs from any Supabase service, just as you would in our cloud Platform.

![Realtime local logs](/images/blog/lw7-o11y/realtime-logs-screenshot.png)

> > 📢 Check out the [new self-hosting docs](http://localhost:3001/docs/reference/self-hosting-analytics/introduction) to get Logflare up and running as your analytics server.
> >
> > It currently supports a BigQuery backend, and we are actively working on supporting more.

Reviewer 2: ```suggestion
> 📢 Check out the [new self-hosting docs](http://localhost:3001/docs/reference/self-hosting-analytics/introduction) to get Logflare up and running as your analytics server. It currently supports a BigQuery backend, and we are actively working on supporting more.
```

side note @MildTomato - do we have an "Callout" component that we could use here? A bit strange as a quote

Reviewer 4: yea, i'll take a look

Reviewer 4: <img width="1262" alt="Screenshot 2023-04-10 at 5 31 47 PM" src="https://user-images.githubusercontent.com/8291514/230875366-352601e7-eab3-4fc0-933d-af8a03f63b2e.png">

changed to above for now.
we should get the same Admonitions setup in both docs and www so you can just do 

```
:::note

Some **content** with _Markdown_ `syntax`. Check [this `api`](#).

:::
```

Reviewer 4: fixed

## The Ingestion Pipeline

> Logflare receives Supabase log events via multiple methods. Services like Postgres use Vector to clean and forward log events to our the Logflare ingest API. Other services such as Realtime and Storage utilize our [integrations](https://github.com/Logflare/logflare#integrations) to send the log events directly. These then get processed and streamed into BigQuery.

Reviewer 2: ```suggestion
Logflare receives Supabase log events via multiple methods. Services like Postgres use [Vector](vector.dev/) to clean and forward log events to our the Logflare ingest API. Other services such as Realtime and Storage utilize our [integrations](https://github.com/Logflare/logflare#integrations) to send the log events directly. These then get processed and streamed into BigQuery.
```

## The Querying Pipeline

> The hard part comes after ingesting the logs: searching, aggregating ,and analyzing them at scale. Crunching many terabytes of data on each query is expensive, and exposing the ingested data to Supabase customers in a naive manner would cause our costs to skyrocket.

Reviewer 2: ```suggestion
After ingesting the logs comes the hard part: searching, aggregating ,and analyzing them at scale. Crunching many terabytes of data on each query is expensive, and exposing the ingested data to Supabase customers in a naive manner would cause our costs to skyrocket.
```

Reviewer 4: @Ziinc for your approvol

To solve these issues, we built and refined Logflare Endpoints, the query engine that powers many of Supabase’s features, such as the logs views, Logs Explorer, and usage charts.

> With Endpoints, you can create HTTP API endpoints from a SQL query, including parameterized queries. Endpoints is like a PostgREST view but with some benefits:

Reviewer 2: ```suggestion
With Endpoints, you can create HTTP API endpoints from a SQL query, including parameterized queries. Endpoints are like PostgREST views but with some benefits:
```

- **Query parameters**
  - You can provide string parameters to the SQL query via the HTTP endpoint.
- **Read-through caching**
  - Results from the query are cached in memory for fast response times.
  - A read-through cache provides results if cached results do not exist.
- **Active cache warming**
  - Query results are proactively warmed at a configurable interval for a combination of fast response times and as-realtime-as-needed data.
- **Query sandboxing**

>   - If an Endpoint query contains a CTE and the sandbox option is selected, the Endpoint will inject the query string of the `sql` query parameter into the Endpoint SQL replacing the default query (the part of the SQL query after the CTE).

Reviewer 2: this one is still very confusing @Ziinc - I don't understand exactly what it means

      "url": "https://api.github.com/users/JohannesBauer97"
    },
    "starred_at": "2023-03-30T20:33:55Z",
		...
  },
  "timestamp": 1680208436849642
}
```

We’re interested in the top contributors, which can be extracted with SQL (in BigQuery dialect):

```sql
select
  count(t.timestamp) as count,
  s.login as gh_user
from
  `github.supabase.webhooks` as t
  cross join unnest(metadata) as m
  cross join unnest(m.sender) as s
where
  timestamp::date > current_date() - @day::int
group by
  gh_user
order by
  count desc
limit
  25
```

With this view in place, we can use Endpoints to provide an API that we can hit from our application:

```bash
curl "https://logflare.app/endpoints/query/69425db0-1cfb-48b4-84c7-2a872b6f0a61" \
 -H 'Content-Type: application/json; charset=utf-8' \
 -G -d "day=30"
```

This returns a JSON response with the top org wide contributors for the last 30 days!

```json
{
  "result": [
    {
      "count": 23404,
      "gh_user": "vercel[bot]"
    },
    {
      "count": 10005,
      "gh_user": "joshenlim"
    },
    {
      "count": 7026,
      "gh_user": "MildTomato"
    },
    {
      "count": 6405,
      "gh_user": "fsansalvadore"
    },

>     {
>       "count": 5195,
>       "gh_user": "saltcod"
>     },
>     {
>       "count": 3454,
>       "gh_user": "alaister"
>     },
>     {
>       "count": 2691,
>       "gh_user": "kevcodez"
>     },
>     {
>       "count": 2117,
>       "gh_user": "gregnr"
>     },
>     {
>       "count": 1769,
>       "gh_user": "Ziinc"
>     },
>     {
>       "count": 1749,
>       "gh_user": "chasers"
>     },
>     {
>       "count": 1430,
>       "gh_user": "Isaiah-Hamilton"
>     },
>     ...

Reviewer 2: ```suggestion
    ...
```

Reviewer 2: if you really want to leave these in then squash them:

```
  "result": [
    { "count": 23404,  "gh_user": "vercel[bot]" },
    { "count": 10005, "gh_user": "joshenlim"  },
    
    etc
```

Reviewer 4: fxied

Reviewer 4: fixed **
